tools/comdb2ar/increment.cpp

```diff
@@ -0,0 +1,523 @@
+        if(!read_octal_ull(head.h.size, sizeof(head.h.size), filesize)) {
+            throw Error("Bad block: bad size");
+        }
+
+        bool is_manifest = false;
+        if(filename == "INCR_MANIFEST") {
+            is_manifest = true;
+        }
+
+        std::string ext;
+        size_t dot_pos = filename.find_first_of('.');
+        if(dot_pos != std::string::npos) {
+            ext = filename.substr(dot_pos + 1);
+        }
+
+        bool is_incr_data = false;
+        bool is_data_file = false;
+        bool is_queue_file = false;
+        bool is_queuedb_file = false;
+
+        if(ext == "data") {
+            is_incr_data = true;
+        } else if (ext == "sha") {
+            sha_fingerprint = read_serialised_sha_file();
+            std::clog << "Fingerprint: " << sha_fingerprint << std::endl;
+            continue;
+        } else {
+            // If it's not a text file then look at the extension to see if it looks
+            // like a data file.  If it does then add it to our list of tables if
+            // not already present.  I was going to trust the lrl file for this
+            // (and use the dbname_file_vers_map for llmeta dbs) but the old
+            // comdb2backup script doesn't serialise the file_vers_map, so I can't
+            // do that yet. This way the onus is on the serialising side to get
+            // the list of files right.
+            if(filename.find_first_of('/') == std::string::npos) {
+                std::string table_name;
+
+                if(recognise_data_file(filename, false, is_data_file,
+                            is_queue_file, is_queuedb_file, table_name) ||
+                   recognise_data_file(filename, true, is_data_file,
+                            is_queue_file, is_queuedb_file, table_name)) {
+                    if(table_set.insert(table_name).second) {
+                        std::clog << "Discovered table " << table_name
+                            << " from data file " << filename << std::endl;
+                    }
+                }
+            }
+        }
+
+        if(is_manifest) {
+
+            // Manifest is first file in each increment, so if one is seen,
+            // delete previous logs if necessary
+            if(!keep_all_logs) {
+                std::clog << "Deleting old log files" << std::endl;
+                clear_log_folder(datadestdir, dbname);
+            }
+
+            manifest_read = true;
+            std::string manifest_sha;
+            std::string manifest_text = read_incr_manifest(filesize);
+            if(!process_incr_manifest(manifest_text, datadestdir,
+                    updated_files, new_files, deleted_files,
+                    file_order, options, manifest_sha)){
+                // Failed to read a coherant manifest
+                std::ostringstream ss;
+                ss << "Error reading manifest";
+                throw Error(ss);
+            }
+
+            if(manifest_sha != sha_fingerprint){
```

Comment:
process_incr_manifest should be there now, for some reason the file was caught in .gitignore

the sha is written as a string in the manifest, all SHA's are pulled from the same timestring